### PR TITLE
Test the rest of the worker protocol serializers

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

We want to test as much of the protocol as possible!

# Context

#8923 created the framework and a few initial tests; this tests the rest of the serializers.

Part of the `BuildResult` test is commented out because we have caught a roundtrip bug! A future PR will fix the bug and uncomment that test.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libstore/build-result.cc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#include "build-result.hh"
+
+namespace nix {
+
+GENERATE_CMP_EXT(
```
Comment (thufschmitt) [inline src/libstore/build-result.cc:5]: Not something for this PR, but the documentation for this macro is quite awful. It took me a while to understand what it was doing (despite having more or less written the first version of it)

File: src/libstore/tests/worker-protocol.cc [hunk 2]
```diff
@@ -136,4 +137,190 @@ CHARACTERIZATION_TEST(
...
+                            .outPath = StorePath { "g1w7hy3qg1w7hy3qg1w7hy3qg1w7hy3q-bar" },
+                        },
+                    },
+                },
+                .startTime = 30,
+                .stopTime = 50,
+                .cpuUser = std::chrono::milliseconds(500s),
+                .cpuSystem = std::chrono::milliseconds(604s),
+            },
+#endif
```
[the remark this answers is not in the document]
Comment (thufschmitt) [reply src/libstore/tests/worker-protocol.cc:234]: Why is is commented out?

[the remark this answers is not in the document]
Comment (thufschmitt) [reply src/libstore/tests/worker-protocol.cc:234]: Aaah. Can you just add a comment to it mentioning that so that we don't forget?

**Review by thufschmitt — APPROVED** [review]
Feel free to merge once https://github.com/NixOS/nix/pull/9094#discussion_r1346349447 is addressed

**Comment by Ericson2314** [comment]
Good idea; done; thanks!

**Pull request merged**